### PR TITLE
Make "libdeno_isolate" interior mutable

### PR DESCRIPTION
**Pull request by F001**
Remove the `UserDataScope` type.

As discussed in https://github.com/denoland/deno/pull/1279, @ry suggested to remove this workaround. The type `UserDataScope` was introduced to prevent memory corruption. We'd better solve the root cause of the corruption.

The solution is that we make sure everything is cleaned up in `DenoIsolate`'s destructor. 

I didn't remove the `user_data` parameter from `deno_execute` and `deno_respond`, because I think that is a good thing.

### CODE REVIEW

File: libdeno/api.cc [hunk 1]
```diff
@@ -187,9 +188,4 @@ void deno_delete(Deno* d_) {
   deno::DenoIsolate* d = reinterpret_cast<deno::DenoIsolate*>(d_);
   delete d;
 }
-
-void deno_terminate_execution(Deno* d_) {
-  deno::DenoIsolate* d = reinterpret_cast<deno::DenoIsolate*>(d_);
-  d->isolate_->TerminateExecution();
-}
```
Comment (ry) [inline libdeno/api.cc]: Why remove this? Please keep.

File: libdeno/deno.h [hunk 2]
```diff
@@ -60,10 +60,6 @@ int deno_execute(Deno* d, void* user_data, const char* js_filename,
 // If this is called after deno_recv_cb has returned, the deno_respond
 // will call into the JS callback specified by libdeno.recv().
 //
-// (Ideally, but not currently: After calling deno_respond(), the caller no
-// longer owns `buf` and must not use it; deno_respond() is responsible for
-// releasing its memory.)
-//
```
Comment (ry) [inline libdeno/deno.h]: Why remove this comment? This is still the aim - unless I missed something we're still memcpy'ing the response.

File: libdeno/internal.h [hunk 3]
```diff
@@ -37,6 +39,16 @@ class DenoIsolate {
...
+    array_buffer_allocator_ = nullptr;
+    isolate_ = nullptr;
+    shared_ = empty_buf;
+    current_args_ = nullptr;
+    snapshot_creator_ = nullptr;
+    global_import_buf_ptr_ = nullptr;
+    pending_promise_events_ = 0;
+    recv_cb_ = nullptr;
+    next_req_id_ = 0;
+    user_data_ = nullptr;
```
Comment (ry) [inline libdeno/internal.h]: Is this useful?

File: libdeno/internal.h [hunk 4]
```diff
@@ -64,24 +76,6 @@ class DenoIsolate {
...
-    CHECK(deno->user_data_ == nullptr || deno->user_data_ == data_);
-    prev_data_ = deno->user_data_;
-    deno->user_data_ = data;
-  }
-
-  ~UserDataScope() {
-    CHECK(deno_->user_data_ == data_);
-    deno_->user_data_ = prev_data_;
-  }
-};
```
Comment (ry) [inline libdeno/internal.h]: Generally I'm ok with this - we ought to address the root cause of the issue.

However @piscisaureus debugged this issue - so he I defer approval to him. Bert, when reviewing this have a look at https://github.com/denoland/deno/commit/60c008d23b2bdad333711b43148a5053e83a62cc which may have addressed the original problem.

File: libdeno/internal.h (hunk 3 above)
Comment (piscisaureus) [reply libdeno/internal.h]: I don't think this is necessary -- but if we keep it: it's too much code. memset() would be better.

Comment (ry) [reply libdeno/internal.h]: I think neither is useful. This is the purpose of ASAN.

That said, if @F001 wants this for debugging, with the expectation that we will remove it later (when ASAN runs on deno) that's okay with me. I prefer this to memset.

**Comment by piscisaureus** [comment]
@ry, @F001 

I think that we can indeed do without the the UserDataScope class.

It is important that C functions that that are called by Rust "borrow" a
reference to the rust Isolate if they're going to (indirectly) mutate it. 

Otherwise we'll re-introduce the bug that was fixed by the commit that
introduced this UserDataScope thing (e742af10a).

It's probably best if the C function stores the rs_isolate ~reference~pointer somewhere.
Otherwise  theoretically maybe LLVM could see through our scheme.

A refresher: the bug, which happened only in Release builds, was something 
like this (simplified):

```rust
pub struct Isolate {
  c_isolate_ptr: *mut OpaqueStruct;
  ncount: u32;
  // ...other fields omitted.
}

fn event_loop(rs_isolate: &mut Isolate) {
  loop {
    unsafe { handle_event_in_c_and_js(rs_isolate.c_isolate_ptr) }; 
    // ^^^ This call may change ncount. But Rust has no idea! It thinks the
    // current scope contains a mutable ref to `rs_isolate`, and nothing seems
    // to happen with it. Since there cannot legally be another mutable ref to
    // `rs_isolate`, there's no way the `ncount` field could be changing.
   
    if (rs_isolate.ncount == 0) { break; }
    // ^^^^ `ncount` does not get read from the struct, because it is assumed
    // that it wasn't changed by `handle_event_in_c()`.
    // (Note that in this example the conditional could be optimized away entirely,
    // but IIRC that wasn't happening -- it was just not reloaded.)
  }
}

// Rust function gets called by C function below.
pub fn do_something_in_rust(c_isolate_ptr: *mut OpaqueStruct) {
  let rs_isolate = unsafe { deno_get_rs_isolate(c_isolate_ptr) as ??? as &mut Isolate };
  // ^-- A second mutable ref to the rust isolate is made here out of thin air.
  // (PS: I simplified the cast a little, please go easy on me.)
  
  rs_isolate->ncount--;
}
```

```c
struct Isolate {
  void* rs_isolate_ptr;
  /* ...other fields left out. */
}

/* C function that gets called from Rust. */
void handle_event_in_c_and_js(void* c_isolate_ptr); {
  struct Isolate* c_isolate = (struct Isolate*) c_isolate_ptr;
  /* Do some stuff which in the end calls back into Rust. */
  do_something_in_rust(c_isolate_ptr);
}

/* C function to retrieve the raw pointer to the Isolate data. */
void* deno_get_rs_isolate(void* c_isolate_ptr) {
   struct Isolate* c_isolate = (struct Isolate*) c_isolate_ptr;
   return c_isolate->rs_isolate_ptr;
}
```

The solution was to let `handle_event_in_c_and_js()` "borrow" the reference to the
rust Isolate. That had no purpose for the C function itself, since it can simply
get a pointer to the the rust Isolate by looking in `cbut it'd let Rust
know that the `rs_isolate` was on the loose and could be mutated.

**Comment by F001** [comment]
@piscisaureus According to your description, it looks like this problem should be solved by [UnsafeCell](https://doc.rust-lang.org/beta/std/cell/struct.UnsafeCell.html). Copy the document here for reference:

> If you have a reference &SomeStruct, then normally in Rust all fields of SomeStruct are immutable. The compiler makes optimizations based on the knowledge that &T is not mutably aliased or mutated, and that &mut T is unique. UnsafeCell<T> is the only core language feature to work around this restriction. All other types that allow internal mutability, such as Cell<T> and RefCell<T>, use UnsafeCell to wrap their internal data.

Recently, I have changed all the `Isolate`'s methods to accept a shared reference `&` in [this PR](https://github.com/denoland/deno/pull/1279).
I think it is reasonable to wrap the `libdeno_isolate` field in a `Cell<T>`, 

```rust
pub struct Isolate {
  libdeno_isolate: Cell<*const libdeno::isolate>,
  ...
}
```
because `libdeno_isolate` members could be modified through below method:
```rust
fn execute(&self,  js_filename: &str, js_source: &str)
```

This will let the compiler know that whenever we use `libdeno_isolate` field, it might have been changed.

**Comment by piscisaureus** [comment]
@F001 Thanks a lot! The `Cell` solution looks sound to me. 

Does using Cell make it possible to remove the `void* user_data` argument from `deno_execute()` and `deno_respond()`, and go back to setting `user_data` just once in `deno_new()` ?

**Comment by F001** [comment]
@piscisaureus I think that is a separate issue, which depends on whether the location of `Isolate` will be changed after the initialization.
I have created another issue https://github.com/denoland/deno/issues/1284 , please feel free to comment.
If we save the `user_data` only in `deno_new()`, we should guarantee that the `Isolate` object will not be moved, which means below usage might be buggy:
```rust
let isolate = Isolate::new(...); // the location of isolate has been saved in C
let boxed_isolate = Box::new(isolate);
// use boxed_isolate below
// We will encounter invalidate pointer issue
```
There is no safe way to express this constraint in Rust as far as I know.
I would prefer to pass `user_data` in `deno_execute()` and `deno_respond()`. Of course the `Isolate` still can not move because the interaction between C and Rust are asynchronous. But at least, we can allow users to move the object after construction before any other method is called.

**Comment by F001** [comment]
Or, we introduce another API `set_user_data()`, and use `std::sync::Once` in `deno_execute()` and `deno_respond()` to make sure the `user_data` is set exactly once. This way, we can also avoid to pass the same `user_data` every time.
But calling `set_user_data()` only in `deno_new()` is still not an option to me.

**Comment by piscisaureus** [comment]
@F001 Ok, you beat me again by answering my question before I asked it. Thanks :)

**Comment by piscisaureus** [comment]
@F001 Although I still wonder why libdeno_isolate now needs to be in a Cell.
Afaik there is no rust code that reads/writes libdeno fields directly.

**Comment by F001** [comment]
To be honest, this is only my intuition. I can't prove it is correct or not. I'm eager to know how to reproduce the problem. If it is impractical to do so, I'd suggest to involve more experienced rust developers for suggestions.

**Comment by piscisaureus** [comment]
@F001 
I don't think we need to put the reference to libdeno::isolate in a Cell.
If we ever get mysterious bugs where the c++ isolate seems to get corrupted I'll try it.
That said, thanks for the effort, we really need your Rust expertise.

**Pull request closed without merging**